src/test/groovy/com/example/demo/BookingTests.groovy

```diff
@@ -0,0 +1,416 @@
+import spock.lang.Specification
+
+import java.time.LocalDate
+
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
+
+@ContextConfiguration
+@SpringBootTest(classes = DemoApplication.class)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+class BookingTests extends Specification {
+
+    MockMvc mockMvc
+
+    @Autowired LocationRepository locationRepository
+    @Autowired CustomerRepository customerRepository
+    @Autowired CarRepository carRepository
+    @Autowired WebApplicationContext context
+    @Autowired BookingRepository bookingRepository
+
+    def setup() {
+        this.mockMvc = MockMvcBuilders.webAppContextSetup(context).build()
+    }
+
+    def "BookingEntity test"() {
+        given: "an entity"
+        def name = "Gary"
+        def customer = new CustomerEntity(name)
+        customerRepository.save(customer)
+
+        def entity = new LocationEntity("Belfast")
+        locationRepository.save(entity)
+
+        def name_car = "Test Car"
+        def reg = "12345"
+        def manufacturer = "Ford"
+        def model = "Fiesta"
+        def transmission = Transmission.MANUAL
+        def category = Category.DIESEL
+        def car = new CarEntity(name_car, reg, manufacturer, model, transmission, category, entity)
+        when: "BOOKING is created with given specifications"
+        def start = LocalDate.now()
+        def end = start.plusDays(1)
+        def booking = new BookingEntity(start, end, car, customer)
+        then: "entity attributes match getter"
+        booking.getId() == 0
+        booking.getStartDate() == start
+        booking.getEndDate() == end
+        booking.getCustomer().getName() == customer.getName()
+        booking.getCar().getRegistration() == car.getRegistration()
+    }
+
+    def "BookingController - REST - adding valid booking test"() {
+        given: "an entity"
+        def entity = new LocationEntity("Cork")
+        locationRepository.save(entity)
+
+        def name = "Gary"
+        def customer = new CustomerEntity(name)
+        customerRepository.save(customer)
+
+        def car_name = "Test Car"
+        def reg = "12345"
+        def manufacturer = "Ford"
+        def model = "Fiesta"
+        def transmission = Transmission.MANUAL
+        def category = Category.DIESEL
+        def car = new CarEntity(car_name, reg, manufacturer, model, transmission, category, entity)
+        carRepository.save(car)
+
+        def booking_json = '{"startDate":"2019-12-17","endDate":"2019-12-18","car":{"id":1,"name":"Test Car","registration":"12345","manufacturer":"Ford","model":"Fiesta","transmission":"MANUAL","category":"DIESEL","location":{"id":1,"location":"Cork"}},"customer":{"id":1,"name":"Gary"}}'
```

Comment:
as Json can be a pain to write, typos etc, you can use `JsonOutput.toJson()` to convert a map/list etc into json, which means you can see you haven't missed out a quote or extra comma etc